File: advanced-artifactory-plugin/src/main/java/leon/patmore/AdvancedArtifactoryProject.java
package leon.patmore;

import org.gradle.api.Project;
import org.gradle.api.publish.maven.plugins.MavenPublishPlugin;
import org.jfrog.gradle.plugin.artifactory.ArtifactoryPlugin;
import org.jfrog.gradle.plugin.artifactory.ArtifactoryPluginUtil;
import org.jfrog.gradle.plugin.artifactory.dsl.ArtifactoryPluginConvention;
import org.jfrog.gradle.plugin.artifactory.dsl.PublisherConfig;
import org.slf4j.Logger;

import java.text.SimpleDateFormat;
import java.util.Date;

public class AdvancedArtifactoryProject {

    private static final SimpleDateFormat VERSION_DATE_FORMAT = new SimpleDateFormat("yyyyMMdd.HHmmss");

    private final Project project;
    private final Logger logger;

    public AdvancedArtifactoryProject(Project project) {
        this.project = project;
        this.logger = project.getLogger();
    }

    void init() {
        logger.info("Loading advanced artifactory plugin!");
        String version = getVersion();
        logger.info("Setting version [ {} ]", version);
        project.setVersion(version);

        applyArtifactoryPlugins();
        setupArtifactoryRepo();
    }

    private void setupArtifactoryRepo() {
        ArtifactoryPluginConvention pluginConvention = ArtifactoryPluginUtil.getArtifactoryConvention(project);
        PublisherConfig publisherConfig = new PublisherConfig(pluginConvention);
        pluginConvention.setPublisherConfig(publisherConfig);
        pluginConvention.setContextUrl(getExpectedEnvVar("ARTIFACTORY_URL"));
        pluginConvention.getPublisherConfig().repository(repo -> {
            repo.setUsername(getExpectedEnvVar("ARTIFACTORY_USERNAME"));
            repo.setPassword(getExpectedEnvVar("ARTIFACTORY_PASSWORD"));
            repo.setRepoKey(getExpectedEnvVar("ARTIFACTORY_REPO_KEY"));
        });
        pluginConvention.getPublisherConfig().defaults(task -> {
            task.publications("mavenJava");
            task.setPublishArtifacts(true);
        });
    }

    private void applyArtifactoryPlugins() {
        project.getPluginManager().apply(MavenPublishPlugin.class);
        project.getPluginManager().apply(ArtifactoryPlugin.class);
    }

    private String getVersion() {
        String gitBranchEnvVar = getExpectedEnvVar("GIT_BRANCH", "undef");
        String[] branchSplit = gitBranchEnvVar.split("/");
        String branchName = branchSplit.length > 1 ? branchSplit[1] : branchSplit[0];
        String commitHash = getExpectedEnvVar("GIT_COMMIT", "undef");
        String shortHash = commitHash.length() > 7 ? commitHash.substring(0, 7) : commitHash;
        return VERSION_DATE_FORMAT.format(new Date()) + "_" + branchName + "_" + shortHash;
    }

    private String getExpectedEnvVar(String name) {
        return getExpectedEnvVar(name, "");
    }

    private String getExpectedEnvVar(String name, String defaultValue) {
        String val = System.getenv(name);
        if (val == null) {
            logger.warn("Expected env var {} is not present!", name);
            return defaultValue;
        }
        return val;
    }

}

File: advanced-artifactory-plugin-kotlin/src/main/kotlin/leon/patmore/AdvancedArtifactoryProject.kt
package leon.patmore

import org.gradle.api.Project
import org.gradle.api.publish.PublishingExtension
import org.gradle.api.publish.maven.MavenPublication
import org.jfrog.gradle.plugin.artifactory.dsl.ArtifactoryPluginConvention
import org.jfrog.gradle.plugin.artifactory.dsl.PublisherConfig
import org.jfrog.gradle.plugin.artifactory.task.ArtifactoryTask

class AdvancedArtifactoryProject(private val project: Project) {

    private val logger = project.logger

    fun init() {
        setRepositories()
        applyPlugins()
        setJavaPublication()
        configureArtifactoryPlugin()
    }

    private fun setRepositories(){
        project.repositories.mavenCentral()
    }

    private fun applyPlugins() {
        project.pluginManager.apply("java")
        project.pluginManager.apply("java-library")
        project.pluginManager.apply("maven-publish")
        project.pluginManager.apply("com.jfrog.artifactory")
    }

    private fun setJavaPublication() {
        val publishing = project.extensions.getByType(PublishingExtension::class.java)
        publishing.publications {
            it.create("mavenJava", MavenPublication::class.java) { publication ->
                (publication as MavenPublication).from(project.components.findByName("java"))
            }
        }
    }

    private fun configureArtifactoryPlugin() {
        val artifactory = project.convention.findPlugin(ArtifactoryPluginConvention::class.java)
        artifactory!!.setContextUrl(getExpectedEnvVar("ARTIFACTORY_URL"))

        artifactory.publish {
            val config = PublisherConfig(artifactory)
            config.setPublishPom(true)
            config.setPublishIvy(true)
            config.repository {
                it.setRepoKey(getExpectedEnvVar("ARTIFACTORY_REPO_KEY"))
                it.setUsername(getExpectedEnvVar("ARTIFACTORY_USERNAME"))
                it.setPassword(getExpectedEnvVar("ARTIFACTORY_PASSWORD"))
            }
        }

        val task = project.tasks.getByName("artifactoryPublish") as ArtifactoryTask
        task.publications("mavenJava")
    }

    private fun getExpectedEnvVar(name: String, defaultValue: String = "") : String{
        val value = System.getenv(name)
        if (value == null) {
            logger.warn("Expected env var {} is not present!", name)
            return defaultValue
        }
        logger.info("Got env var $value for $name")
        return value
    }

}
